### ISSUE TITLE
Periodic `RESCAN` to update mounted drives

### ISSUE THREAD

**Issue by Burrito78** [dosbox-staging/dosbox-staging#1265]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Mac OS

### Is your feature request related to a problem? Please describe.

Users are encouraged to manage their files on the host system. 
Also tools to effectively manage files inside Dosbox are missing and are out of scope (deltree, file manager etc.).
I regularly can't remember the name of the 'rescan' command or forget that CTRL+F4 does also work.

-> I change my files on the host system and can't see the changes inside Dosbox automatically. This is frustrating.

### Describe alternatives you've considered

Closing and reopening Dosbox (always do this, because its the least work).

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by nemo93** [comment]
I'm adding a note to the wiki about this. Also I've found out a potential workaround thanks to @ripsaw to be read [there](https://www.vogons.org/viewtopic.php?t=46942).

**Comment by arrowgent** [comment]
press ctrl+f4 to update drives

**Comment by Burrito78** [comment]
Hi @arrowgent, thanks for your feedback. I already know how to do it manually but this ticket is about a possible automation of `RESCAN`.

**Comment by arrowgent** [comment]
personally i dont see automation as a benefit

while i do agree ive had to tell people on the internet to "update drives" by pressing ctrl+f4
this seems more like a tutorial/lack of knowledge about program operation

then we get into how often is appropriate for automatically rescanning?
is automation required for many games/software?
do people OFTEN edit host files while running a live dosbox? (yes i do, and i press ctrl+f4)
should it be an option in dosbox conf? enable/disable
does the option offer a MS delay?  1000ms?  what if you need 50ms rescan periods? 
would setting this to 1ms effectively lockup dosbox?  should there be set limits 10ms-1000ms?
does the timing configuration effect performance?


also since there is a workaround using `z:\rescan.com` or` -t floppy`, does adding _yet another_ feature like this have significance?

again, personally, i dont like automation tricks especially if they take away user's control.

**Comment by kcgen** [comment]
Once populated, the drive cache keeps DOSBox in memory and avoid big context switches out to the system libraries (virus scanners catching on-read events), and adding physical disk-latency into DOSBox's 1-ms tick loop.

Of course, disk IO is necessary though, and so those "hits" are taken whenever the cache doesn't have those records.

The problem is, caches are meant to invalidate themselves.   Image a world where all caches require manual intervention to invalidate caches -- where the consequence (if not manually triggered) is that old data records are used? 
 - CPU L1 to L3 cache: auto-invalidated on source memory-page changes
 - block device caches: auto-invalidated on source block changes
 - IO read cache: auto-invalidated on file or directory changes
 - Web-server rendered-output cache: auto-invalidated on source-asset (db records, static content, sources) changes
 - Database query caches: auto-invalidated on source rol/column content changes
 - to CDROM read caches: never invalidated due to read only (only memory pressure eviction)

I could keep going. 

Bottom line is: caches are meant to be auto-invalidated.  A world where people had to manually intervene would be a world where everything fails due to people failing at this task - and therefore caches would be abandoned.

Can the DOSBox cache be auto-invalidated? Probably.. this isn't a black art or something.

But perhaps in this case (for users tinkering about on the command line), it will be easier to offer a "disable cache" setting.  But the gold-standard will be auto-invalidation.

**Comment by Burrito78** [comment]
Thanks @kcgen for the through explanation! In my case, i'm on the DOSBox command line while i decide to copy install files, disk images or a game directory in on the host system. I personally wouldn't need the `RESCAN` automation while running a emulated game. I don't see a possible implementation of that feature to take user control away but more like removing unnecessary administrative burden for end users.

I want DOSBox to work in a way that someone would expect it to who had a PC in the 90s but never used DOSBox before. He should be able to get a old game running without the help of Google. :)

**Comment by kcgen** [comment]
> I want DOSBox to work in a way that someone would expect it to who had a PC in the 90s but never used DOSBox before

:+1: 

If we can auto-invalidate its cache, then this will get us closer to how `smartdrv` behaved.

**Comment by arrowgent** [comment]
i was thinking also how that would effect CD emulation while reading the "iso" or physical disk
would pressing ctrl+f4 (rescan) constantly effect this?

momentarily you could create a TSR RESCAN .com file and run it in your [autoexec] section
although a TSR is more designed to modify memory space than run a command repeatedly

**Comment by Burrito78** [comment]
@kcgen Is this maybe a Backlog candidate?

**Comment by GranMinigun** [comment]
Perhaps updating caches on regained window focus would be enough.

**Comment by Burrito78** [comment]
@kcgen 

I would be happy if the update would happen every 3-5 seconds. 
If that's not practical then it's time to close this issue.

**Comment by joncampbell123** [comment]
The DOSBox-X MOUNT command has a -nocachedir option which was added to disable the cache for the drive letter you mount. Would that help?

As usual, Staging is free to borrow that. src/dos/dos_programs.cpp and src/dos/drive_local.cpp

**Comment by johnnovak** [comment]
Yeah I was wondering why do we even need a cache. WinUAE works fine without a cache when mounting host folders as harddrives.

**Comment by Grounded0** [comment]
This -nocachedir sounds good to me. It could even be our default setting if people can't come up with downsides against it.

**Comment by johnnovak** [comment]
Yep, would be good to understand why the caching was introduced in the first place. What specific problems is it supposed to solve, and are those problems still there and/or relevant?

**Comment by joncampbell123** [comment]
It depends on the OS because the emulator has to stat() the file on the host to determine how to present it to the guest.

On Linux, stat() is very fast because of the efficient VFS layer in the kernel, while on Windows, stat() and filesystem I/O in general is quite sluggish because of code deep within the NT kernel that nobody wants or knows how to optimize. Yes, even with SSDs.

To give you an idea how slow it is... when I used to make MinGW builds of DOSBox-X on my Windows 10 laptop, it took 4-5 *hours* to compile all those variations because the sluggish I/O system in Windows absolutely made MinGW crawl (you'll notice Git is quite slow on Windows too). Even more bizarre, I found that MinGW gets much better performance on that same system if you bring up VirtualBox and a VM with Windows XP and compile MinGW inside that VM. Yes, on the same Windows 10 system you get better performance in a VM than on the actual filesystem :laughing:  And this is on a laptop who's hard drive I long ago replaced with an SSD.

**Comment by johnnovak** [comment]
Thanks @joncampbell123, that's informative.

I've also experiences general slowness when using Git on Windows, but Git usage patterns that creates all those millions of little files in the `.git` subfolder is not quite comparable to the contents of a typical DOS directory with maybe 10-20 files on average. Definitely below 100! So I guess we'll need to test and see how this behaves with realistic DOS HDD contents, not millions of little files. My guess is it will be just fine, plus Win 10/11 is not exactly Win 2000 either 😄

**Comment by johnnovak** [comment]
For the record, I ran into some issues with re-scanning the mounted drive while using some programs (QPV, in this case). The problem is that as DOS was not a multi-tasking operating system, programs can make assumptions that the contents of the filesystem will not change under them. E.g. they might cache the contents of a directory listing, etc. I don't really understand the specific reasons, but I ended up with a lot of weird behaviour when I kept copying files into the mounted folder on the host side, then used the file management functions of QPV after rescanning the mounted drive (but without quitting QPV). Ultimately I got it to hang/crash...

So I still think it's worthwhile to give this feature a go, but it should be off by default, and I'm pretty sure it can only be used reliable with known good programs that don't misbehave if the contents of the filesystem change under them without notice.

**Comment by shermp** [comment]
> Thanks @joncampbell123, that's informative.
> 
> I've also experienced general slowness when using Git on Windows, but Git usage patterns that creates all those millions of little files in the `.git` subfolder is not quite comparable to the contents of a typical DOS directory with maybe 10-20 files on average. Definitely below 100! So I guess we'll need to test and see how this behaves with realistic DOS HDD contents, not millions of little files. My guess is it will be just fine, plus Win 10/11 is not exactly Win 2000 either 😄
> 
> I'm not quite sold on the concept that slow stat calls alone could cause such long compilation times on MinGW, maybe it contributes to it, but there must be something else going on. It seems very unlikely if you think about it; it cannot be _that_ slow, and it cannot be called _that_ many times that the slowdown would add up to hours. Maybe general MinGW slowness is more like it as Linux software was simply never optimised for Windows.
> 
> Also, from all this follows that this issue is most likely non-existent on macOS.

I'm pretty sure the big part of the issue with MinGW on Windows is automake generated makefiles. I remember when doing build stuff for PCem that the handmade makefiles were waaaaay faster than the automake ones created in MSYS2

**Comment by Torinde** [comment]
As for SSDs - while my host PC is with SSD, I'm very often using DOSbox from a SMB share on a HDD.

**Comment by johnnovak** [comment]
I think we should simply disable the cache and see what happens on Linux, Windows, and macOS. If there is no performance penalty on current OSes (remember, Win 10 is the min target), the FS cache goes to the chopping board...

Alternatively, we can adopt Jon's `nocache` option. See his above comment for implementation details.

**Comment by FeralChild64** [comment]
I would prefer to eventually get rid of the cache, if only it is feasible. The cache adds another layer of complexity.

**Comment by weirddan455** [comment]
There's a lot to dive in to in this ticket.  First of all, let me make it clear the cache is *not* storing actual file data.  The OS does a good job of that.  On `localDrive` open, read, and write call `fopen` `fread` `fwrite` directly every time.

The cache mainly deals with storing file names and mapping them from host name to DOS name.  This is an essential part of file operations and is non-trivial.  `drive_cache.cpp` is over 1000 lines of C-style code so even refactoring it to re-scan the host file system on every operation would be non-trivial.

There could very well be a large performance hit by scanning the host file system on every operation as well.  Linux and Mac have case-sensitive filesystems whereas DOS is case-insensitive.  This means for a file open, you potentially have to scan the host directory for every permutation of casing.

What @joncampbell123 said about Windows file I/O performance is also true... although that gets a bit complicated.  Programs written for Linux are often doing file I/O in an inefficient matter but they are able to get away with it because calls like `stat()` are much cheaper on Linux than Windows.

I'd recommend watching this video if you want to learn more.  It's by a developer of the Rust installer and what he had to do to make it performant on Windows:  https://www.youtube.com/watch?v=qbKGw8MQ0i8

I don't think Windows overhead would be our biggest problem though.  In fact, Windows may end up performing better because NTFS is case-insensitive like DOS.  The biggest performance problem would likely be the overhead of DOS to host naming and the biggest development problem would be refactoring `drive_cache.cpp`

Another possible solution is using host side notification of file system changes to update the cache.  Linux has `inotify` (https://man7.org/linux/man-pages/man7/inotify.7.html) and I'm sure Windows has something similar.

Again though, just refactoring `drive_cache.cpp` is a pretty large job and is currently a critical piece of our file I/O system.

**Comment by johnnovak** [comment]
@weirddan455 Yeah I agree with pretty much all you wrote above. The cache must stay as we need a way to map between host filesystem and DOS filenames.

`inotify` and the equivalent on Windows on macOS is the real solution, yeah; getting rid of the cache is not an option IMO.

Also, we need the cache to handle high-ASCII characters properly. I raised a ticket for it a while ago, and while it's conceptually simple, I looked into this briefly today, and man... what a huge steaming mess all the file I/O in DOSBox is 🥶 💩 This is ripe for a grand refactor before anything new is inserted into it; it's almost undecipherable what it does and how. 

So here's my ticket in question:
https://github.com/dosbox-staging/dosbox-staging/issues/2619

> Again though, just refactoring drive_cache.cpp is a pretty large job and is currently a critical piece of our file I/O system.

Yeah, and a piecewise refactoring won't work here, unfortunately, as it's so messy, full of C-isms (string handling character by character... ugh), mutation everywhere, functions doing a hundred things and doing random side effects, globals, piles of ifdefs, casual C file I/O functions here and there (despite there is a cross platform I/O layer too), and you name it...

**Comment by weirddan455** [comment]
I had a look at what DOSBox-X was doing.  All their "nocache" mount option does is set a flag on that drive to call `EmptyCache()` before every function that would look into it.  The rescan command also just calls `EmptyCache()` on the current drive and the CTRL+F4 hotkey (meant for swapping between multiple CDROM or floppy images) calls `EmptyCache()` on _every_ drive.

Implementing a "nocache" mount option in this way would be pretty simple.  It should certainly be off by default though due to both performance and correctness concerns.  This could be a quick fix if it works "well enough" for people that need this feature.

>  For the record, I ran into some issues with re-scanning the mounted drive while using some programs (QPV, in this case).

As @johnnovak said, I'm not surprised there's bugs.  `drive_cache.cpp` as a whole probably needs a good review and refactor but that's a lot of work :sweat_smile: 

>  Perhaps updating caches on regained window focus would be enough.

This would be too heavy handed.  It would be easy enough to write code to call `EmptyCache()` on an SDL focus event but I'm not sure that's a good idea.  If we did it, it would need to be behind an off by default config option.

Porting over X's mount option is the quick fix but I would not be surprised if you encounter subtle (or not so subtle) bugs while using it.

**Comment by johnnovak** [comment]
@weirddan455 I don't like the X approach, too heavy handed, yeah. I don't like the Alt-Tab approach either.

What's wrong with `inotify` that you brought up? Can't we do that, and the macOS and Windows equivalents, so just invalidate on actual FS changes?

**Comment by weirddan455** [comment]
>  What's wrong with inotify that you brought up? Can't we do that, and the macOS and Windows equivalents, so just invalidate on actual FS changes?

It's just more work to do that and this is not very high on my priority list.  Probably would be a fine approach though.

**Comment by johnnovak** [comment]
> > What's wrong with inotify that you brought up? Can't we do that, and the macOS and Windows equivalents, so just invalidate on actual FS changes?
> 
> It's just more work to do that and this is not very high on my priority list. Probably would be a fine approach though.

Yeah, I'd say either `inotify`, or just leave the manual Ctrl+F4 refresh in place for now.

**Comment by rderooy** [comment]
I think `inotify` is a good solution, just keep it mind it works only for local filesystems. e.g., it will not work with network filesystems such as NFS. May want to mention that somewhere in the docs, as you know someone will ask about that at some point.

**Comment by joncampbell123** [comment]
There are two APIs in Linux, the more recent can tell you about files that are deleted, created, etc. and you only need one object to modify files in the directory and it's subdirectories.

The older one:
https://www.man7.org/linux/man-pages/man7/inotify.7.html

The newer one (Linux 5.1 and newer):
https://www.man7.org/linux/man-pages/man7/fanotify.7.html

**Comment by joncampbell123** [comment]
Equivalent Windows API, I think... https://learn.microsoft.com/en-us/windows/win32/fileio/obtaining-directory-change-notifications

**Comment by joncampbell123** [comment]
I think this is the equivalent API on Mac OS, not sure: https://developer.apple.com/documentation/coreservices/file_system_events?language=objc

**Comment by johnnovak** [comment]
Thanks @joncampbell123, very informative 👍🏻

**Comment by johnnovak** [comment]
@weirddan455 I'm thinking of closing this with "out of scope". Too much hassle for a very niche edge case (messing around with the mounted directory on the host side, and wanting for the changes to show up in the same DOSBox session). This is not really needed for gaming, and you can always just hit the rescan hotkey manually.

**Comment by johnnovak** [comment]
Too complicated for little benefit for gaming.
